core/src/Cuda/Kokkos_Cuda_Instance.cpp

```diff
@@ -851,7 +848,31 @@ Cuda::Cuda(cudaStream_t stream, Impl::ManageStream manage_stream)
       }) {
   Impl::CudaInternal::singleton().verify_is_initialized(
       "Cuda instance constructor");
-  m_space_instance->initialize(stream, static_cast<bool>(manage_stream));
+  m_space_instance->initialize(Impl::CudaInternal::singleton().m_cudaDev,
+                               stream, static_cast<bool>(manage_stream));
+}
+
+Cuda::Cuda(int device_id, cudaStream_t stream)
```

Comment:
Is there a way to ensure that `stream` was created on `device_id` (I'm not seeing anything obvious in CudaAPI)? I'm guessing it could cause issues if they are mismatched.

Would it make sense that this constructor only taking the device id and internally creating a stream where we can ensure it is on the correct device?

Reply:
We discussed that we don't want to (allow to) manage the stream if we don't use the default execution space instance.

Reply:
> Is there a way to ensure that stream was created on device_id (I'm not seeing anything obvious in CudaAPI)? I'm guessing it could cause issues if they are mismatched.

We could possibly use (not tested)
```
CUresult cuStreamGetCtx ( CUstream hStream, CUcontext* pctx );
CUresult cuCtxPushCurrent ( CUcontext ctx )
CUresult cuCtxGetDevice ( CUdevice* device )
CUresult cuCtxPopCurrent ( CUcontext* ctx )
```
but I'm not sure if it's worth it.

Reply:
@tcclevenger I added such a check in e156d5859e36fbc3c4f84575a0c8734689c1b3eb.